op-node/rollup/derive/attributes_queue.go

```diff
@@ -58,82 +48,43 @@ func (aq *AttributesQueue) Step(ctx context.Context, outer Progress) error {
-	return io.EOF
-}
-
-func (aq *AttributesQueue) SafeL2Head() eth.L2BlockRef {
-	return aq.next.SafeL2Head()
-}
-
-// DeriveL2Inputs turns the next L2 batch into an Payload Attributes that builds off of the safe head
-func (aq *AttributesQueue) DeriveL2Inputs(ctx context.Context, l2SafeHead eth.L2BlockRef) (*eth.PayloadAttributes, error) {
 	if len(aq.batches) == 0 {
-		return nil, io.EOF
+		return io.EOF
 	}
 	batch := aq.batches[0]
 
-	seqNumber := l2SafeHead.SequenceNumber + 1
-	// Check if we need to advance an epoch & update local state
-	if l2SafeHead.L1Origin != batch.Epoch() {
-		aq.log.Info("advancing epoch in the attributes queue", "l2SafeHead", l2SafeHead, "l2SafeHead_origin", l2SafeHead.L1Origin, "batch_timestamp", batch.Timestamp, "batch_epoch", batch.Epoch())
-		seqNumber = 0
-	}
-
 	fetchCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
 	defer cancel()
-	l1Info, _, receipts, err := aq.dl.Fetch(fetchCtx, batch.EpochHash)
+	attrs, crit, err := PreparePayloadAttributes(fetchCtx, aq.config, aq.dl, aq.next.SafeL2Head(), batch.Epoch())
 	if err != nil {
-		aq.log.Error("failed to fetch L1 block info", "l1Origin", batch.Epoch(), "err", err)
-		return nil, err
-	}
-
-	// Fill in deposits if we are the first block of the epoch
-	var deposits []hexutil.Bytes
-	if seqNumber == 0 {
-		var errs []error
-		deposits, errs = DeriveDeposits(receipts, aq.config.DepositContractAddress)
-		for _, err := range errs {
-			aq.log.Error("Failed to derive a deposit", "l1Origin", batch.Epoch(), "err", err)
-		}
-		if len(errs) != 0 {
-			// TODO: Multierror here
-			return nil, fmt.Errorf("failed to derive some deposits: %v", errs)
+		if crit {
+			return fmt.Errorf("failed to prepare payload attributes for batch: %v", err)
+		} else {
+			aq.log.Error("temporarily failing to prepare payload attributes for batch", "err", err)
+			return nil
 		}
 	}
 
-	var txns []eth.Data
-	l1InfoTx, err := L1InfoDepositBytes(seqNumber, l1Info)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create l1InfoTx: %w", err)
-	}
-	txns = append(txns, l1InfoTx)
-	if seqNumber == 0 {
-		txns = append(txns, deposits...)
-	}
-	txns = append(txns, batch.Transactions...)
-	attrs := &eth.PayloadAttributes{
-		Timestamp:             hexutil.Uint64(batch.Timestamp),
-		PrevRandao:            eth.Bytes32(l1Info.MixDigest()),
-		SuggestedFeeRecipient: aq.config.FeeRecipientAddress,
-		Transactions:          txns,
-		// we are verifying, not sequencing, we've got all transactions and do not pull from the tx-pool
-		// (that would make the block derivation non-deterministic)
-		NoTxPool: true,
-	}
-	aq.log.Info("generated attributes in payload queue", "tx_count", len(txns), "timestamp", batch.Timestamp)
+	// we are verifying, not sequencing, we've got all transactions and do not pull from the tx-pool
+	// (that would make the block derivation non-deterministic)
+	attrs.NoTxPool = true
+	attrs.Transactions = append(attrs.Transactions, batch.Transactions...)
```

Comment:
Manually modifying the attributes to add the batch transactions is a weird interface.

Reply:
I'm thinking of it as preparing a template, building the default empty block attributes. It's then up to the user to modify the template. Do you have a suggestion for a better interface, without duplicating code?

Reply:
after reading it a bit more, I think I'm fine with this pattern as long as it's documented in the `Prepare` function